### ISSUE TITLE
AGBSAVE options missing on one agbsave.bin of a game.

### ISSUE THREAD

**Issue by krystal-starfox11111** [d0k3/GodMode9#412]
I don't know if this issue is known, but when trying to inject the save for Super Mario Advance 3: Yoshi's Island it doesn't show "AGBSAVE options" to allow it. Based on some old info I've tried dumping the agbsave.bin, then extracting the save by deleting the first 512B, and the rest after the 8KB save size. That file does work in an emulator after being qword byte swapped, but going the other way with my other save doesn't work on 3DS. It still keeps the original file on the 3DS, even if GM9 says it was successfully copied. It is an EEPROM_v122 save. I read there were some issues with those, though that was a while ago and I don't know if you've resolved them. Though, unless I'm missing a checksum I don't know about, manually copying it should be working, right?

**Comment by d0k3** [comment]
Just to make sure you're doing everything right - can you inject savegames to other VC GBA titles?

**Comment by d0k3** [comment]
Okay, I just tried it, worked without flaw for me.

Just to be sure, first you run the game once on your console, then you boot into GM9. You put that savegame (which is *exactly* 8192 byte) into the clipboard, as the only entry. Then you move to `S:/agbsave.bin`, press A, `AGBSAVE options` -> `Inject savegame`.

If you're still having trouble, extract the current savegame from there and compare it in a hex editor with the one that doesn't work. Also makee sure you're running the newest version of GM9.

**Comment by d0k3** [comment]
I tested with the Ambassador one.

> It just doesn't show AGBSAVE options on that one game.
Just to be clear, AGBSAVE options turn up for `S:/agbsave.bin` and nowhere else.

*If* in case of that specific game you don't see the AGBSAVE options, you will need to copy the `S:/agbsave.bin` file and send that to me.

**Comment by krystal-starfox11111** [comment]
Here's the agbsave.bin from my 3DS.
[agbsave.zip](https://github.com/d0k3/GodMode9/files/2281501/agbsave.zip)

**Comment by d0k3** [comment]
Oh well, I have no clue how you managed to get this into your `agbsave.bin`. It says you saved a total of 2084724354 times (did you play that often?). In any case, this looks like it was seriously messed with in some way or another.

Anyways, here's a test build that should handle this game for you. Test it, give me feedback:
https://f.secretalgorithm.com/kW1cu/godmode9-v1.7.1-4-ge44d266d-20180814002535.zip

Now, i seriously don't want to include that fix in the GM9 source code, cause it will lead to false positives.  I need to be 100% sure there was nothing obvious that caused this. Sooo... can you give me the exact steps (as exact as possible) you do when injecting the save?

Also, there is no trouble with this (same game, same save) on the 2nd 3DS? You can inject it and it works?

**Comment by krystal-starfox11111** [comment]
Yes, it does save the original file from the same system on the 1st 3DS for me. No, I didn't save it anywhere near that many times. I have no idea how my file got like that. I doubt it was the system transfer from launch 3DS to New 3DS XL. Just wondering, is the save count part of the GBA or 3DS portion of the data?

My 1st 3DS (New 3DS XL Majora's Mask version from release in case initial firmware had anything to do with it), it's still 11.7.0-40U, I didn't install B9S nor Luma on this one, hopefully that isn't the issue because all the other saves inject perfectly. 2nd 3DS (small New 3DS Super Mario Black version) v11.4.0-37U, B9S v1.3, Luma v9.1, GodMode9 v1.7.1.

What I did to inject the save with the test version: Start Yoshi's Island, Press Home, close it, when the home menu finishes loading the one of random official themes on shuffle, I press the power button to show the power off menu, then press the power button again to actually power off the 3DS, boot GM9 via NTRBootHax, select 0: SDCARD (3DS XL), go into my folder "1GBA-Saves", Press Y to copy file "SMA 3 - Yoshi's Island.sav", back out twice, go to S: SYSNAND VIRTUAL, press A on agbsave.bin (now it shows AGBSAVE options), select AGBSAVE options, select Inject GBA VC save, says Savegame inject success, press start to reboot, start Yoshi's Island, and it does have the new save, copied the file to another slot, close the game, launch another GBA game, go back to Yoshi, and the save does still have the copied slot.

For version 1.7.1, it was similar so I'll copy and edit from the previous one, and this one is from memory of what I did then: Start Yoshi's Island, Press Home, close it, when the home menu finishes loading the one of random official themes on shuffle, I press the power button to show the power off menu, then press the power button again to actually power off the 3DS, boot GM9 via NTRBootHax, select 0: SDCARD (3DS XL), go into my folder "1GBA-Saves", Press Y to copy file "agbsave.bin" (the one I modified to insert the byte swapped save, I've also copied the working injected agbsave.bin from my 2nd 3DS), back out twice, go to S: SYSNAND VIRTUAL, when I press A on agbsave.bin it doesn't show AGBSAVE options, Calculate CMAC fails, I back out of that, press Y to paste, press A to confirm paste "agbsave.bin" here, the bar flashes by, Calculate CMAC still fails, press start to reboot, start Yoshi's Island, and it still shows the original save not the copied one.

I tried copying the agbsave.bin from my 1st 3DS to my 2nd, but it doesn't work. Copying and injecting over the extracted save from using the test GM9 from 1st to 2nd does work. I'll get to the uninstall/reinstall test later.

BTW this is unrelated, but is it possible to decrypt an old save from a Pokemon Y cartridge played on a Gateway 3DS clone CFW and re-encrypt it to work on current OFW without having that 3DS and Gateway clone? People said it had to do with the Key-Y being different between v4,5 and later versions.

**Comment by d0k3** [comment]
Okay, sorry it took me a few days to get back.

> BTW this is unrelated, but is it possible to decrypt an old save from a Pokemon Y cartridge played on a Gateway 3DS clone CFW and re-encrypt it to work on current OFW without having that 3DS and Gateway clone? People said it had to do with the Key-Y being different between v4.5 and later versions.
I'm not that knowledgeable on that stuff, so take the info I give you with a grain of salt. The relevant keys here should be 0x33 and 0x37 (see [here](https://www.3dbrew.org/wiki/AES_Registers)). Both are not system unique. 0x2F (which is the v6.0) console key isn't system unique either. In short, just inserting the cart and continue playing should work.

Moving it over to a CIA install will require a save manager like JKSM, but should still work. If there is more trouble maybe ask [J-D-K](https://github.com/J-D-K/JKSM)?

> Just wondering, is the save count part of the GBA or 3DS portion of the data?
It's part of the header, which is 3DS stuff. What you got in there just isn't correct. But it seems to work (when not injecting), so that's a riddle to solve.

Sooo... did my test build not fix the issue for you? That one is special, it's not identical with release or current nightly. Just to be safe, here's the link again:
https://f.secretalgorithm.com/kW1cu/godmode9-v1.7.1-4-ge44d266d-20180814002535.zip

**Comment by d0k3** [comment]
Okay, I figured I'll give you some more info.

That's how your borked AGBSAVE looks like:
https://i.imgur.com/1uMEWmp.png

This is how it should look like:
https://i.imgur.com/gmpBgWj.png
The red 0xFF's are actually supposed to be 0xFF in the file (and in the past they always were). The red 0x00 is the save size, which is a lower number than what you got in the borked version. Everything from offset 0x10...0x20 is - in simplified terms - a signature (one that we can forge) for the header.

You see that your AGBSAVE is quite different from the expected result, and the riddle is: how can that still work? Anyways, let me know if you actually tried with the test build. If that also didn't work it will get a lot more complicated.

**Comment by d0k3** [comment]
@krystal-starfox11111 - I need more info. Can you help?

**Comment by krystal-starfox11111** [comment]
Sorry I haven't checked here in a while. I started playing Monster Hunter World after the last time because I found out about an FFXIV event that didn't have a lot of time left, and I haven't even played the game until then, but that's irrelevant.

Yes, the test version works on my main 3DS, and does inject the save successfully, sorry if I wasn't clear the first time. I deleted and re-downloaded the game from the eShop, and now it works with both the test, and the main build v1.7.1. Also I thought I'd test if the Test build works with a game that worked with the original version and it does, at least for F-Zero Maximum Velocity.
